File: libraries/fary/ary_form.f
      SUBROUTINE ARY_FORM( IARY, FORM, STATUS )
*+
*  Name:
*     ARY_FORM

*  Purpose:
*     Obtain the storage form of an array.

*  Language:
*     Starlink Fortran 77

*  Invocation:
*     CALL ARY_FORM( IARY, FORM, STATUS )

*  Description:
*     The routine returns the storage form of an array as an upper-case
*     character string (e.g. 'SIMPLE').

*  Arguments:
*     IARY = INTEGER (Given)
*        Array identifier.
*     FORM = CHARACTER * ( * ) (Returned)
*        Storage form of the array.
*     STATUS = INTEGER (Given and Returned)
*        The global status.

*  Notes:
*     -  The symbolic constant ARY__SZFRM may be used for declaring the
*     length of a character variable to hold the storage form of an
*     array. This constant is defined in the include file ARY_PAR.
*     -  At present, the ARY_ routines only support "primitive", "scaled"
*     and "simple" arrays, so only the values 'PRIMITIVE', 'SCALED' and 
*     'SIMPLE' can be returned.

*  Algorithm:
*     -  Import the array identifier.
*     -  Obtain the Data Control Block index for the data object.
*     -  Ensure that form information is available for the data object.
*     -  Copy the form string to the output argument.
*     -  If an error occurred, then report context information.

*  Copyright:
*     Copyright (C) 1989 Science & Engineering Research Council.
*     All Rights Reserved.
*     Copyright (C) 2006 Particle Physics and Astronomy Research
*     Council. All Rights Reserved.

*  Licence:
*     This program is free software; you can redistribute it and/or
*     modify it under the terms of the GNU General Public License as
*     published by the Free Software Foundation; either version 2 of
*     the License, or (at your option) any later version.
*     
*     This program is distributed in the hope that it will be
*     useful,but WITHOUT ANY WARRANTY; without even the implied
*     warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
*     PURPOSE. See the GNU General Public License for more details.
*     
*     You should have received a copy of the GNU General Public License
*     along with this program; if not, write to the Free Software
*     Foundation, Inc., 59 Temple Place,Suite 330, Boston, MA
*     02111-1307, USA

*  Authors:
*     RFWS: R.F. Warren-Smith (STARLINK)
*     {enter_new_authors_here}

*  History:
*     14-JUN-1989 (RFWS):
*        Original version.
*     {enter_changes_here}

*  Bugs:
*     {note_any_bugs_here}

*-
      
*  Type definitions:
      IMPLICIT NONE              ! No implicit typing

*  Global Constants:
      INCLUDE 'SAE_PAR'          ! Standard SAE constants
      INCLUDE 'DAT_PAR'          ! DAT_ public constants
      INCLUDE 'ARY_PAR'          ! ARY_ public constants
      INCLUDE 'ARY_CONST'        ! ARY_ private constants

*  Global Variables:
      INCLUDE 'ARY_DCB'          ! ARY_ Data Control Block
*        DCB_FRM( ARY_MXDCB ) = CHARACTER * ( ARY__SZFRM ) (Read)
*           Form string for data object.

      INCLUDE 'ARY_ACB'          ! ARY_ Access Control Block
*        ACB_IDCB( ARY__MXACB = INTEGER (Read)
*           Index to data object entry in the DCB.

*  Arguments Given:
      INTEGER IARY

*  Arguments Returned:
      CHARACTER * ( * ) FORM

*  Status:
      INTEGER STATUS             ! Global status

*  Local variables:
      INTEGER IACB               ! Index to ACB entry
      INTEGER IDCB               ! Index to data object entry in the DCB

*.

*  Check inherited global status.
      IF ( STATUS .NE. SAI__OK ) RETURN

*  Import the array indentifier.
      CALL ARY1_IMPID( IARY, IACB, STATUS )
      IF ( STATUS .EQ. SAI__OK ) THEN

*  Get the DCB index for the data object.
         IDCB = ACB_IDCB( IACB )

*  Ensure that form information is available.
         CALL ARY1_DFRM( IDCB, STATUS )

*  Copy the form string to the output argument.
         CALL ARY1_CCPY( DCB_FRM( IDCB ), FORM, STATUS )
      END IF
       
*  If an error occurred, then report context information and call the
*  error tracing routine.
      IF ( STATUS .NE. SAI__OK ) THEN
         CALL ERR_REP( 'ARY_FORM_ERR',
     :   'ARY_FORM: Error determining the storage form of an array.',
     :   STATUS )
         CALL ARY1_TRACE( 'ARY_FORM', STATUS )
      END IF             

      END

File: libraries/cary/aryForm.c
#include "sae_par.h"
#include "ary1.h"
#include "mers.h"
#include "dat_par.h"
#include <string.h>

void aryForm( Ary *ary, char form[ARY__SZFRM+1], int *status ) {
/*
*+
*  Name:
*     aryForm

*  Purpose:
*     Obtain the storage form of an array.

*  Synopsis:
*     void aryForm( Ary *ary, char form[ARY__SZFRM+1], int *status )

*  Description:
*     This function returns the storage form of an array as an upper-case
*     character string (e.g. 'SIMPLE').

*  Parameters:
*     ary
*        Array identifier.
*     form
*        Returned holding the storage form of the array.
*     status
*        The global status.

*  Notes:
*     -  The symbolic constant ARY__SZFRM should be used for declaring the
*     length of a character variable to hold the storage form of an
*     array. This constant is defined in the header file ary.h.
*     -  At present, the ARY_ routines only support "primitive", "scaled",
*     "simple" and "delta" arrays, so only the values 'PRIMITIVE', 'SCALED'
*     'DELTA' and 'SIMPLE' can be returned.

*  Copyright:
*      Copyright (C) 2017 East Asian Observatory
*      All rights reserved.

*  Licence:
*     This program is free software; you can redistribute it and/or
*     modify it under the terms of the GNU General Public License as
*     published by the Free Software Foundation; either version 2 of
*     the License, or (at your option) any later version.
*
*     This program is distributed in the hope that it will be
*     useful,but WITHOUT ANY WARRANTY; without even the implied
*     warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
*     PURPOSE. See the GNU General Public License for more details.
*
*     You should have received a copy of the GNU General Public License
*     along with this program; if not, write to the Free Software
*     Foundation, Inc., 51 Franklin Street,Fifth Floor, Boston, MA
*     02110-1301, USA

*  Authors:
*     RFWS: R.F. Warren-Smith (STARLINK)
*     DSB: David S. Berry (EAO)

*  History:
*     03-JUL-2017 (DSB):
*        Original version, based on equivalent Fortran routine by RFWS.

*-
*/

/* Local variables: */
   AryACB *acb;
   AryDCB *dcb;

/* Check inherited global status. */
   if( *status != SAI__OK ) return;

/* Inport the array identifier. */
   acb = (AryACB *) ary1Impid( ary, 1, 1, 1, status );
   if( *status == SAI__OK ){

/* Get the DCB index for the data object. */
      dcb = acb->dcb;

/* Ensure that form information is available. */
      ary1Dfrm( dcb, status );

/* Copy the form string to the output argument. */
      if( *status == SAI__OK ) strcpy( form, dcb->form );
   }

/* If an error occurred, then report context information and call the error
   tracing routine. */
   if( *status != SAI__OK ){
      errRep( " ", "aryform: Error determining the storage form of an array.",
              status );
      ary1Trace( "aryForm", status );
   }

}
